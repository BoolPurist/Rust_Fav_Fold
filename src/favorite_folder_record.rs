use serde::{Deserialize, Serialize};
use std::path::Path;

use crate::trimmed_not_empty_text::{NonEmptyText, NotEmptyTextError};

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
/// Struct to encapsulate a valid favorite created by the user
/// This favorite is a not empty label with an absolute path
/// That path does need to point to existing location.
pub struct FavoriteFolderPath {
    name: NonEmptyText,
    location: NonEmptyText,
}

impl FavoriteFolderPath {
    /// # Errors
    ///
    /// - If name or path are empty or white spaces.
    /// - If path can not be resolved to a valid utf 8 string.
    /// - If a relative path can not be resolved to an absolute one.
    pub fn new(name: NonEmptyText, location: NonEmptyText) -> Self {
        Self { name, location }
    }

    pub fn set_name(&mut self, new_name: NonEmptyText) {
        self.name = new_name;
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }

    pub fn path_str(&self) -> &str {
        &self.location
    }
    pub fn path(&self) -> &Path {
        Path::new(self.location.as_str())
    }
    pub fn does_exit(&self) -> bool {
        Path::new(self.location.as_str()).exists()
    }
}

/// Error to describe the cause of failing to create a struct [`FavoriteFolderPath`]
use thiserror::Error;
#[derive(Debug, Error)]
pub enum InvalidFavoriteFields {
    #[error("Name of favorite folder should not be empty or only whitespace")]
    EmptyName(#[from] NotEmptyTextError),
    #[error("Path of favorite folder should not be empty or only whitespace")]
    EmptyPath,
    #[error("Path of favorite folder is not a valid utf 8 text")]
    InvalidUtf8PathStr,
}
