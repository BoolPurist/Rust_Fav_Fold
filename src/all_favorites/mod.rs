use std::path::Path;

use crate::favorite_folder_record::FavoriteFolderPath;

#[derive(Debug, PartialEq, Eq)]
pub struct AllFavorites(Vec<FavoriteFolderPath>);

impl From<AllFavorites> for Vec<FavoriteFolderPath> {
    fn from(value: AllFavorites) -> Self {
        value.0
    }
}

impl AllFavorites {
    pub fn new(folders: impl IntoIterator<Item = FavoriteFolderPath>) -> Self {
        Self(folders.into_iter().collect())
    }

    pub fn clean_all_dangling(&mut self, mut on_is_dangling_check: impl FnMut(&Path) -> bool) {
        self.0
            .retain_mut(|maybe_dangling| !on_is_dangling_check(maybe_dangling.path()));
    }
}

#[cfg(test)]
mod testing {
    use std::path::Path;

    use crate::{favorite_folder_record::FavoriteFolderPath, AllFavorites};

    const INPUT: &str = include_str!("test_input.json");
    fn given() -> AllFavorites {
        let parsed: Vec<FavoriteFolderPath> = serde_json::from_str(INPUT).unwrap();
        AllFavorites::new(parsed)
    }
    #[test]
    fn if_dangling_delete_a_path() {
        let expected = AllFavorites::new([FavoriteFolderPath::new(
            "proc_macro",
            &Path::new("/home/some_user/Code/rust/proc-macro-workshop"),
        )
        .unwrap()]);

        let mut given = given();
        given.clean_all_dangling(|path| {
            let path = path.to_str().unwrap();
            path == "/home/some_user/Documents/Studium" || path == "/home/some_user/Code/rust"
        });
        assert_eq!(expected, given);
    }
}
