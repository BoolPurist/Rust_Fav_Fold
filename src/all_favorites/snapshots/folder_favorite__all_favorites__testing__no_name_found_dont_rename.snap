---
source: src/all_favorites/mod.rs
expression: actual
---
AllFavorites(
    [
        FavoriteFolderPath {
            name: NonEmptyText(
                "studium",
            ),
            location: NonEmptyText(
                "/home/some_user/Documents/Studium",
            ),
        },
        FavoriteFolderPath {
            name: NonEmptyText(
                "dev_rust",
            ),
            location: NonEmptyText(
                "/home/some_user/Code/rust",
            ),
        },
        FavoriteFolderPath {
            name: NonEmptyText(
                "proc_macro",
            ),
            location: NonEmptyText(
                "/home/some_user/Code/rust/proc-macro-workshop",
            ),
        },
    ],
)
